packages/flutter_tools/lib/src/dart/pub.dart

```diff
@@ -484,8 +504,9 @@ class _DefaultPub implements Pub {
 
     final String cachePath = _fileSystem.path.join(Cache.flutterRoot!, '.pub-cache');
     if (_fileSystem.directory(cachePath).existsSync()) {
-      _logger.printTrace('Using $cachePath for the pub cache.');
-      return cachePath;
+      joinCaches(_fileSystem, '${_platform.environment['Home']}/.pub_cache', cachePath);
```

Comment:
Logic for finding the default `PUB_CACHE` is here:
https://github.com/dart-lang/pub/blob/73ea9a98b5bef9cd2b1ca4d1eba8b6d20e1ad3d0/lib/src/system_cache.dart#L40-L60

There is occasionally some discussion around moving to a more XDG compliant location. But that's not happening right now. Feel free to file a PR against `pub` with a `//NOTE: ....` saying that when changing this we should also patch `flutter_tools`.